src/dune_rpc_impl/server.ml

```diff
@@ -60,15 +69,48 @@ end)
 
 module Session_set = Session_comparable.Set
 
+module Subscribers = struct
+  type t =
+    { build_progress : Session_set.t
+    ; diagnostics : Session_set.t
+    }
+
+  let empty =
+    { build_progress = Session_set.empty; diagnostics = Session_set.empty }
+
+  let get t (which : Subscribe.t) =
+    match which with
+    | Build_progress -> t.build_progress
+    | Diagnostics -> t.diagnostics
+
+  let modify t (which : Subscribe.t) ~f =
+    match which with
+    | Build_progress -> { t with build_progress = f t.build_progress }
+    | Diagnostics -> { t with diagnostics = f t.diagnostics }
+
+  let add t which session =
+    modify t which ~f:(fun set -> Session_set.add set session)
+
+  let remove t which session =
+    modify t which ~f:(fun set -> Session_set.remove set session)
+
+  let notify t which ~f =
+    let set = get t which in
+    Fiber.parallel_iter_set (module Session_set) set ~f
```

Comment:
Shouldn't all this stuff be the job of the RPC? It seems odd to me that the build system has to care about the list of client that are connected.

Reply:
I guess I don't consider `dune_rpc_impl/server` to be part of the build system. I could move this higher up to `dune_rpc_server`.

Reply:
Well, here we have a property that multiple clients can subscribe to, in order to be notified when it changes. The property in question here is the of current build errors.

Subscribing to a property is a general concept. If we want to declare another property somewhere else in Dune, are we going to copy and paste this code?

It seems to me that the notion of subscribing to a property is general enough that it should be nicely wrapped up in the RPC library, so that in `build_system.ml` all we do is set the new value and be done with it. There is an additional twist here: the value is big so we send delta rather than complete values, but that seems like a general enough concept as well to me.

For instance, in D-Bus there is also a notion of property that can change over time. In the code that implement a property, all you do is declare the property and then call `set` on it. Who should be notified of this `set` is the buisiness of the D-Bus library. I would expect this to be the same with other IPC systems as well.

Reply:
> It seems to me that the notion of subscribing to a property is general enough that it should be nicely wrapped up in the RPC library, so that in build_system.ml all we do is set the new value and be done with it. There is an additional twist here: the value is big so we send delta rather than complete values, but that seems like a general enough concept as well to me.

Adding new types of subscriptions is easy enough. You just add a new field to Subscribers.t and modify `get`.

> For instance, in D-Bus there is also a notion of property that can change over time. In the code that implement a property, all you do is declare the property and then call set on it. Who should be notified of this set is the buisiness of the D-Bus library. I would expect this to be the same with other IPC systems as well.

Yes, that's a good point and it would be much nicer to have signal semantics (signal in the sense of `React.S`). However, I think that's a pretty big change and we should first discuss the API in an issue.